File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:

     // optimize one
       int maxProfit(vector<int>& prices) {
        int n= prices.size();
        int ptr = INT_MAX; 
        int a=0;   
        int b=0;

        for(int i=0; i<n;i++){
           if(prices[i] < ptr){
               ptr = prices[i];
           }
            a = prices[i] - ptr;
            if(b < a){
                b = a;
            }
        }
      
 
    return b;
        
    }
    // Brute force
    // int maxProfit(vector<int>& prices){
    //    int n = prices.size();
    //    int ans = INT_MIN;

    //    for(int i=0; i<n-1; i++){
    //       for(int j=i+1; j<n; j++){
    //         ans = max(ans, prices[j] - prices[i]);
    //       }
    //    }
    //    if(ans <0)return 0;
    //    else return ans;
    // }

   
  
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.go
func maxProfit(prices []int) int {
    
// Brute force
    // n := len(prices)
    // var result int
    // for i:=0 ; i<n-1; i++{
    //     for j:= i+1; j<n; j++{
    //         if prices[i] < prices[j]{
    //             profit := prices[j]-prices[i]
    //             if profit > result{
    //                 result = profit
    //             }
    //         }
    //     }      
    // }
    // return result

 // Optimizes Solution 
   minPrice := prices[0]
   maxProfit := 0

   for i:=1; i<len(prices); i++{
    if prices[i] < minPrice {
        minPrice = prices[i]
    }else if prices[i]- minPrice > maxProfit{
        maxProfit = prices[i] - minPrice
    }   
   }  
   return maxProfit
}